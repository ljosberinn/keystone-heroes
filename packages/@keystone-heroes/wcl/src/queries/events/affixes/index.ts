import { Affixes } from "@keystone-heroes/db/types";

import type { EventParams } from "..";
import type {
  AbsorbEvent,
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyDebuffStackEvent,
  DamageEvent,
  HealEvent,
  InterruptEvent,
} from "../types";
import {
  filterExpression as bolsteringFilterExpression,
  getBolsteringEvents,
} from "./bolstering";
import {
  filterExpression as burstingFilterExpression,
  getBurstingEvents,
} from "./bursting";
import {
  filterExpression as explosiveFilterExpression,
  getExplosiveEvents,
} from "./explosive";
import {
  filterExpression as grievousFilterExpression,
  getGrievousEvents,
} from "./grievous";
import {
  filterExpression as necroticFilterExpression,
  getNecroticEvents,
} from "./necrotic";
import {
  filterExpression as quakingFilterExpression,
  getQuakingEvents,
} from "./quaking";
import {
  filterExpression as sanguineFilterExpression,
  getSanguineEvents,
} from "./sanguine";
import {
  filterExpression as spitefulFilterExpression,
  getSpitefulEvents,
} from "./spiteful";
import {
  filterExpression as stormingFilterExpression,
  getStormingEvents,
} from "./storming";
import {
  filterExpression as tormentedFilterExpression,
  getTormentedEvents,
} from "./tormented";
import {
  filterExpression as volcanicFilterExpression,
  getVolcanicEvents,
} from "./volcanic";

type AffixWithEvents = Exclude<
  Affixes,
  | "Awakened"
  | "Beguiling"
  | "Tyrannical"
  | "Fortified"
  | "Prideful"
  | "Raging"
  | "Infested"
  | "Inspiring"
  | "Reaping"
  | "Skittish"
>;

const affixExpressionMap: Record<AffixWithEvents, string[]> = {
  [Affixes.Sanguine]: sanguineFilterExpression,
  [Affixes.Explosive]: explosiveFilterExpression,
  [Affixes.Grievous]: grievousFilterExpression,
  [Affixes.Necrotic]: necroticFilterExpression,
  [Affixes.Volcanic]: volcanicFilterExpression,
  [Affixes.Bursting]: burstingFilterExpression,
  [Affixes.Spiteful]: spitefulFilterExpression,
  [Affixes.Quaking]: quakingFilterExpression,
  [Affixes.Storming]: stormingFilterExpression,
  [Affixes.Bolstering]: bolsteringFilterExpression,
  [Affixes.Tormented]: tormentedFilterExpression,
};

export const getAffixExpression = (affixes: Affixes[]): string[] => {
  return affixes
    .filter((affix): affix is AffixWithEvents => affix in affixExpressionMap)
    .flatMap((affix) => affixExpressionMap[affix]);
};

export const filterAffixEvents = (
  allEvents: AllTrackedEventTypes[],
  affixes: EventParams["affixes"]
): (
  | AbsorbEvent
  | InterruptEvent
  | HealEvent
  | DamageEvent
  | ApplyBuffEvent
  | ApplyDebuffStackEvent
)[] => {
  const affixSet = new Set(affixes);

  // seasonal
  const tormentedEvents = getTormentedEvents(allEvents, affixSet);

  // common affixes
  const storming = getStormingEvents(allEvents, affixSet);
  const spiteful = getSpitefulEvents(allEvents, affixSet);
  const sanguine = getSanguineEvents(allEvents, affixSet);
  const volcanic = getVolcanicEvents(allEvents, affixSet);
  const quaking = getQuakingEvents(allEvents, affixSet);
  const bolstering = getBolsteringEvents(allEvents, affixSet);
  const bursting = getBurstingEvents(allEvents, affixSet);
  const explosive = getExplosiveEvents(allEvents, affixSet);
  const grievous = getGrievousEvents(allEvents, affixSet);
  const necrotic = getNecroticEvents(allEvents, affixSet);

  return [
    ...tormentedEvents,
    ...storming,
    ...spiteful,
    ...sanguine,
    ...volcanic,
    ...quaking,
    ...bolstering,
    ...bursting,
    ...explosive,
    ...grievous,
    ...necrotic,
  ];
};
