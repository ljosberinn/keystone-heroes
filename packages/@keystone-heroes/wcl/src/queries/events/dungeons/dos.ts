import type { AllTrackedEventTypes, ApplyDebuffEvent } from "../types";
import { createIsSpecificEvent } from "../utils";

export const DOS_URN = 228_626;

/**
 * @see https://www.warcraftlogs.com/reports/hFj3wLzrapC4KvZk#fight=2&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24type%20%3D%20%22applydebuff%22%20and%20target.type%20%3D%20%22NPC%22%20and%20ability.id%20%3D%20228626
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "hFj3wLzrapC4KvZk") {
 *       fights(fightIDs: [2]) {
 *         startTime
 *         endTime
 *       }
 *        events(startTime: 34007, endTime: 1759795, filterExpression: "type = \"applydebuff\" and target.type = \"npc\" and ability.id = 228626") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "applydebuff" and target.type = "npc" and ability.id = ${DOS_URN}`,
];

const isDosUrnEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: DOS_URN,
});

export const getDOSEvents = (
  allEvents: AllTrackedEventTypes[]
): ApplyDebuffEvent[] => {
  return allEvents.filter(isDosUrnEvent);
};
