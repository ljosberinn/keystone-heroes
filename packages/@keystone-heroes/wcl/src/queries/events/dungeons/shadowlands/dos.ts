import { EventDataType, HostilityType } from "../../../../types";
import type {
  ApplyDebuffEvent,
  RefreshDebuffEvent,
  RemoveDebuffEvent,
} from "../../types";
import type { GetEventBaseParams } from "../../utils";
import { getEvents } from "../../utils";

export const DOS_URN = 228_626;

export const getDeOtherSideUrnUsage = async (
  params: GetEventBaseParams
): Promise<ApplyDebuffEvent[]> => {
  const allEvents = await getEvents<
    ApplyDebuffEvent | RemoveDebuffEvent | RefreshDebuffEvent
  >({
    ...params,
    dataType: EventDataType.Debuffs,
    hostilityType: HostilityType.Enemies,
    abilityID: DOS_URN,
  });

  return allEvents.filter(
    (event): event is ApplyDebuffEvent => event.type === "applydebuff"
  );
};
