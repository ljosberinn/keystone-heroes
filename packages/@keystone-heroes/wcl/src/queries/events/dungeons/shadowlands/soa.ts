import { EventDataType, HostilityType } from "../../../../types";
import type {
  ApplyDebuffEvent,
  RemoveDebuffEvent,
  RefreshDebuffEvent,
} from "../../types";
import type { GetEventBaseParams } from "../../utils";
import { createChunkByThresholdReducer, getEvents } from "../../utils";

export const SOA_SPEAR = 339_917;

// SoA Spear stun lasts 10 seconds
// each usage should be thus at least 10s apart of each other
const reducer = createChunkByThresholdReducer(10 * 1000);

export const getSpiresOfAscensionSpearUsage = async (
  params: GetEventBaseParams
): Promise<ApplyDebuffEvent[]> => {
  const allEvents = await getEvents<
    ApplyDebuffEvent | RemoveDebuffEvent | RefreshDebuffEvent
  >({
    ...params,
    dataType: EventDataType.Debuffs,
    hostilityType: HostilityType.Enemies,
    abilityID: SOA_SPEAR,
  });

  // picks the first event of each chunk
  return allEvents
    .filter((event): event is ApplyDebuffEvent => event.type === "applydebuff")
    .reduce<ApplyDebuffEvent[][]>(reducer, [])
    .flatMap((chunk) => chunk[0]);
};
