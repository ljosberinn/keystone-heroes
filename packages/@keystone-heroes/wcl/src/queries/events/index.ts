import type { DungeonIDs } from "@keystone-heroes/db/data/dungeons";
import type { PlayableClass, Affixes } from "@keystone-heroes/db/types";

import {
  deathFilterExpression,
  filterEnemyDeathEvents,
  filterPlayerDeathEvents,
  filterProfessionEvents,
  filterRemarkableSpellEvents,
  remarkableSpellFilterExpression,
} from "../events/other";
import type {
  AllTrackedEventTypes,
  BeginCastEvent,
  DeathEvent,
} from "../events/types";
import { filterAffixEvents, getAffixExpression } from "./affixes";
import { getDungeonExpression, filterDungeonEvents } from "./dungeons";
import { recursiveGetEvents } from "./utils";

export type EventParams = {
  reportID: string;
  startTime: number;
  endTime: number;
  dungeonID: DungeonIDs;
  fightID: number;
  affixes: Affixes[];
};

const generateFilterExpression = ({
  dungeonID,
  affixes,
}: Pick<EventParams, "dungeonID" | "affixes">) => {
  return [
    deathFilterExpression,
    remarkableSpellFilterExpression,
    ...getDungeonExpression(dungeonID),
    ...getAffixExpression(affixes),
  ]
    .flat()
    .map((part) => `(${part})`)
    .join(" or ");
};

export const getEvents = async (
  params: EventParams,
  playerMetaInformation: { actorID: number; class: PlayableClass }[]
): Promise<{
  allEvents: AllTrackedEventTypes[];
  playerDeathEvents: DeathEvent[];
  enemyDeathEvents: (DeathEvent | BeginCastEvent)[];
}> => {
  const filterExpression = generateFilterExpression({
    dungeonID: params.dungeonID,
    affixes: params.affixes,
  });

  const allEvents = await recursiveGetEvents<AllTrackedEventTypes>({
    startTime: params.startTime,
    endTime: params.endTime,
    reportID: params.reportID,
    filterExpression,
  });

  const actorIDSet = new Set(
    playerMetaInformation.map((dataset) => dataset.actorID)
  );

  const dungeonEvents = filterDungeonEvents(
    allEvents,
    params.dungeonID,
    actorIDSet
  );
  const affixEvents = filterAffixEvents(allEvents, params.affixes);
  const professionEvents = filterProfessionEvents(allEvents);
  const remarkableSpellEvents = filterRemarkableSpellEvents(allEvents);
  const playerDeathEvents = filterPlayerDeathEvents(
    allEvents,
    actorIDSet
    // remarkableSpellEvents
  );
  const enemyDeathEvents = filterEnemyDeathEvents(allEvents, actorIDSet);

  return {
    allEvents: [
      ...professionEvents,
      ...dungeonEvents,
      ...affixEvents,
      ...remarkableSpellEvents,
      ...playerDeathEvents,
    ].sort((a, b) => a.timestamp - b.timestamp),
    playerDeathEvents,
    enemyDeathEvents,
  };
};
