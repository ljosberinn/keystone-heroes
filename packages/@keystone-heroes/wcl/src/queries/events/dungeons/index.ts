import { DungeonIDs } from "@keystone-heroes/db/data";
import type { PlayableClass } from "@keystone-heroes/db/types";

import type {
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyBuffStackEvent,
  ApplyDebuffEvent,
  BeginCastEvent,
  CastEvent,
  DamageEvent,
  DeathEvent,
  HealEvent,
  RemoveBuffEvent,
} from "../../events/types";
import { filterExpression as dosFilterExpression, getDOSEvents } from "./dos";
import { filterExpression as hoaFilterExpression, getHOAEvents } from "./hoa";
import { filterExpression as nwFilterExpression, getNWEvents } from "./nw";
import { filterExpression as pfFilterExpression, getPFEvents } from "./pf";
import { filterExpression as sdFilterExpression, getSDEvents } from "./sd";
import { filterExpression as soaFilterExpression, getSOAEvents } from "./soa";
import { filterExpression as topFilterExpression, getTOPEvents } from "./top";

type DungeonWithEvents = Exclude<DungeonIDs, DungeonIDs.MISTS_OF_TIRNA_SCITHE>;

const isDungeonWithEvent = (id: DungeonIDs): id is DungeonWithEvents =>
  id in dungeonExpressionMap;

const dungeonExpressionMap: Record<DungeonWithEvents, string[]> = {
  [DungeonIDs.DE_OTHER_SIDE]: dosFilterExpression,
  [DungeonIDs.HALLS_OF_ATONEMENT]: hoaFilterExpression,
  [DungeonIDs.PLAGUEFALL]: pfFilterExpression,
  [DungeonIDs.SANGUINE_DEPTHS]: sdFilterExpression,
  [DungeonIDs.SPIRES_OF_ASCENSION]: soaFilterExpression,
  [DungeonIDs.THEATRE_OF_PAIN]: topFilterExpression,
  [DungeonIDs.THE_NECROTIC_WAKE]: nwFilterExpression,
};

export const getDungeonExpression = (id: DungeonIDs): string[] =>
  isDungeonWithEvent(id) ? dungeonExpressionMap[id] : [];

export const filterDungeonEvents = (
  allEvents: AllTrackedEventTypes,
  dungeonID: DungeonIDs,
  playerMetaInformation: { actorID: number; class: PlayableClass }[]
):
  | ApplyDebuffEvent[]
  | CastEvent[]
  | (DeathEvent | DamageEvent | ApplyBuffEvent)[]
  | (BeginCastEvent | ApplyBuffEvent | RemoveBuffEvent | ApplyBuffStackEvent)[]
  | (DamageEvent | HealEvent)[] => {
  switch (dungeonID) {
    case DungeonIDs.DE_OTHER_SIDE:
      return getDOSEvents(allEvents);
    case DungeonIDs.HALLS_OF_ATONEMENT:
      return getHOAEvents(allEvents);
    case DungeonIDs.PLAGUEFALL:
      return getPFEvents(allEvents, playerMetaInformation);
    case DungeonIDs.SANGUINE_DEPTHS:
      return getSDEvents(allEvents);
    case DungeonIDs.SPIRES_OF_ASCENSION:
      return getSOAEvents(allEvents);
    case DungeonIDs.THEATRE_OF_PAIN:
      return getTOPEvents(allEvents);
    case DungeonIDs.THE_NECROTIC_WAKE:
      return getNWEvents(allEvents);
    default:
      return [];
  }
};
