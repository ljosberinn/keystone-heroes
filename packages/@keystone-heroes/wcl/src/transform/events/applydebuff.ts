import { EventType } from "@keystone-heroes/db/types";

import { DOS_URN } from "../../queries/events/dungeons/dos";
import { ENVELOPMENT_OF_MISTS } from "../../queries/events/dungeons/mots";
import { SOA_SPEAR } from "../../queries/events/dungeons/soa";
import type { ApplyDebuffEvent } from "../../queries/events/types";
import type { Processor } from "../utils";

const ids = new Set<number>([SOA_SPEAR, DOS_URN, ENVELOPMENT_OF_MISTS]);

/**
 * track SoA Spear & DOS Urn
 */
export const applyDebuffProcessor: Processor<ApplyDebuffEvent> = (event) => {
  if (ids.has(event.abilityGameID)) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.ApplyDebuff,
      abilityID: event.abilityGameID,
    };
  }

  return null;
};
