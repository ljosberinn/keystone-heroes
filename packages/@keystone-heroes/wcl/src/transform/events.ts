import { EventType } from "@keystone-heroes/db/types";
import type { Prisma, Pull, PullZone } from "@keystone-heroes/db/types";
import type { DeepNonNullable } from "ts-essentials";

import { INVISIBILITY } from "../queries";
import { BOLSTERING } from "../queries/events/affixes/bolstering";
import { BURSTING } from "../queries/events/affixes/bursting";
import { EXPLOSIVE } from "../queries/events/affixes/explosive";
import { GRIEVOUS_WOUND } from "../queries/events/affixes/grievous";
import { NECROTIC } from "../queries/events/affixes/necrotic";
import { QUAKING } from "../queries/events/affixes/quaking";
import {
  SANGUINE_ICHOR_DAMAGE,
  SANGUINE_ICHOR_HEALING,
} from "../queries/events/affixes/sanguine";
import { STORMING } from "../queries/events/affixes/storming";
import { VOLCANIC } from "../queries/events/affixes/volcanic";
import { DOS_URN } from "../queries/events/dungeons/shadowlands/dos";
import { NW } from "../queries/events/dungeons/shadowlands/nw";
import {
  PF_GREEN_BUFF,
  PF_RED_BUFF,
  PF_PURPLE_BUFF,
} from "../queries/events/dungeons/shadowlands/pf";
import {
  SD_LANTERN_BUFF,
  SD_LANTERN_OPENING,
} from "../queries/events/dungeons/shadowlands/sd";
import { SOA_SPEAR } from "../queries/events/dungeons/shadowlands/soa";
import { TOP_BANNER_AURA } from "../queries/events/dungeons/shadowlands/top";
import type {
  AnyEvent,
  ApplyBuffEvent,
  ApplyBuffStackEvent,
  ApplyDebuffEvent,
  BeginCastEvent,
  CastEvent,
  DamageEvent,
  DeathEvent,
  HealEvent,
  InterruptEvent,
  RemoveBuffEvent,
} from "../queries/events/types";
import type { ReportMapBoundingBox, ReportDungeonPullNpc } from "../types";

type Processor<T extends AnyEvent, AdditionalParams = Record<string, null>> = (
  event: T,
  params: {
    sourcePlayerID: number | null;
    targetPlayerID: number | null;
    sourceNPCID: number | null;
    targetNPCID: number | null;
  } & AdditionalParams
) => Prisma.EventCreateManyPullInput | null;

/**
 * track invis potion, bolstering, ToP and PF buffs
 */
const applyBuffProcessor: Processor<ApplyBuffEvent & { stacks?: number }> = (
  event,
  { sourcePlayerID, targetPlayerID, targetNPCID }
) => {
  if (sourcePlayerID && targetPlayerID) {
    if (
      sourcePlayerID === targetPlayerID &&
      (event.abilityGameID === INVISIBILITY.POTION_OF_THE_HIDDEN_SPIRIT ||
        event.abilityGameID === INVISIBILITY.DIMENSIONAL_SHIFTER)
    ) {
      return {
        timestamp: event.timestamp,
        sourcePlayerID,
        abilityID: event.abilityGameID,
        eventType: EventType.ApplyBuff,
      };
    }

    if (
      event.abilityGameID === TOP_BANNER_AURA ||
      event.abilityGameID === PF_GREEN_BUFF.aura ||
      event.abilityGameID === PF_RED_BUFF.aura ||
      event.abilityGameID === PF_PURPLE_BUFF.aura
    ) {
      return {
        timestamp: event.timestamp,
        eventType: EventType.ApplyBuff,
        abilityID: event.abilityGameID,
        sourcePlayerID,
        targetPlayerID,
      };
    }
  }

  if (event.abilityGameID === BOLSTERING && targetNPCID) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.ApplyBuff,
      stacks: event.stacks ?? null,
      targetNPCID,
      targetNPCInstance: event.targetInstance ?? null,
    };
  }

  if (event.abilityGameID === SD_LANTERN_BUFF && targetPlayerID) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.ApplyBuff,
      targetPlayerID,
      abilityID: event.abilityGameID,
    };
  }

  return null;
};

/**
 * track SD Lantern usage
 */
const applyBuffStackProcessor: Processor<ApplyBuffStackEvent> = (
  event,
  { targetPlayerID }
) => {
  if (targetPlayerID && event.abilityGameID === SD_LANTERN_BUFF) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.ApplyBuffStack,
      abilityID: event.abilityGameID,
      targetPlayerID,
      stack: event.stack,
    };
  }

  return null;
};

/**
 * track SoA Spear & DOS Urn
 */
const applyDebuffProcessor: Processor<ApplyDebuffEvent> = (event) => {
  if (event.abilityGameID === SOA_SPEAR || event.abilityGameID === DOS_URN) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.ApplyDebuff,
      abilityID: event.abilityGameID,
    };
  }

  return null;
};

/**
 * track SD Lantern usage
 */
const beginCastProcessor: Processor<BeginCastEvent> = (
  event,
  { sourcePlayerID }
) => {
  if (sourcePlayerID && event.abilityGameID === SD_LANTERN_OPENING) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.BeginCast,
      abilityID: event.abilityGameID,
      sourcePlayerID,
    };
  }

  return null;
};

/**
 * track any cast from players
 * - pre-filtered to only contain 1min+ cooldowns
 * - dungeon-specific abilities (HoA Gargoyles)
 */
const castProcessor: Processor<CastEvent> = (
  event,
  { sourcePlayerID, targetNPCID, targetPlayerID }
) => {
  if (sourcePlayerID) {
    return {
      timestamp: event.timestamp,
      abilityID: event.abilityGameID,
      eventType: EventType.Cast,
      targetNPCID,
      targetPlayerID,
      sourcePlayerID,
    };
  }

  return null;
};

const isEnvironmentalDamageAffixAbilityID = (id: number) =>
  id === VOLCANIC ||
  id === STORMING ||
  id === EXPLOSIVE.ability ||
  id === GRIEVOUS_WOUND ||
  id === SANGUINE_ICHOR_DAMAGE ||
  id === BURSTING ||
  id === QUAKING ||
  id === NECROTIC;

/**
 * track any damage taken from
 * - affixes
 *
 * track any player damage done to
 * - explosives (pre-filtered to only contain lasthits)
 */
const damageProcessor: Processor<DamageEvent> = (
  event,
  { targetPlayerID, sourcePlayerID, sourceNPCID, targetNPCID }
) => {
  if (event.amount === 0) {
    return null;
  }

  // player taking damage
  if (targetPlayerID) {
    if (isEnvironmentalDamageAffixAbilityID(event.abilityGameID)) {
      return {
        timestamp: event.timestamp,
        abilityID: event.abilityGameID,
        eventType: EventType.DamageTaken,
        damage: event.amount + (event.absorbed ?? 0),
        targetPlayerID,
        sourcePlayerID,
      };
    }

    // NPC damaging player
    if (sourceNPCID) {
      return {
        timestamp: event.timestamp,
        eventType: EventType.DamageTaken,
        damage: event.amount + (event.absorbed ?? 0),
        targetPlayerID,
        sourceNPCID,
        // only relevant on plagueborer
        abilityID: event.abilityGameID,
      };
    }
  }

  // Player damaging NPC
  if (targetNPCID && sourcePlayerID) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.DamageDone,
      // damage vs Explosives is irrelevant
      damage:
        targetNPCID === EXPLOSIVE.unit
          ? null
          : event.amount + (event.absorbed ?? 0),
      targetNPCID,
      sourcePlayerID,
    };
  }

  return null;
};

/**
 * track deaths:
 * - all player deaths
 * - PF slimes
 */
const deathProcessor: Processor<DeathEvent, { pull: PersistedDungeonPull }> = (
  event,
  { targetNPCID, targetPlayerID, pull }
) => {
  const sourceNPCID = event.killerID
    ? pull.enemyNPCs.find((npc) => npc.id === event.killerID)?.gameID ?? null
    : null;

  // player death
  if (targetPlayerID) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.Death,
      sourceNPCInstance: event.killerInstance ?? null,
      targetPlayerID,
      sourceNPCID,
    };
  }

  // prideful, PF slimes
  return {
    timestamp: event.timestamp,
    eventType: EventType.Death,
    targetNPCInstance: event.targetInstance ?? null,
    targetNPCID,
  };
};

/**
 * track any heal:
 * - sanguine
 * - kyrian orb
 */
const healProcessor: Processor<HealEvent> = (
  event,
  { targetNPCID, targetPlayerID, sourcePlayerID }
) => {
  if (event.amount === 0) {
    return null;
  }

  if (
    event.abilityGameID === NW.KYRIAN_ORB_HEAL &&
    sourcePlayerID &&
    targetPlayerID
  ) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.HealingDone,
      abilityID: NW.KYRIAN_ORB_HEAL,
      healingDone: event.amount,
      sourcePlayerID,
      targetPlayerID,
    };
  }

  if (event.abilityGameID === SANGUINE_ICHOR_HEALING && targetNPCID) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.HealingDone,
      abilityID: SANGUINE_ICHOR_HEALING,
      healingDone: event.amount,
      targetNPCID,
    };
  }

  return null;
};

/**
 * track any interrupts through quaking including the interrupted ability
 */
const interruptProcessor: Processor<InterruptEvent> = (
  event,
  { sourcePlayerID, targetPlayerID }
) => {
  if (sourcePlayerID && targetPlayerID) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.Interrupt,
      interruptedAbilityID: event.extraAbilityGameID,
      sourcePlayerID,
      targetPlayerID,
    };
  }

  return null;
};

/**
 * track SD Lantern usage
 */
const removeBuffProcessor: Processor<RemoveBuffEvent> = (
  event,
  { targetPlayerID }
) => {
  if (event.abilityGameID === SD_LANTERN_BUFF && targetPlayerID) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.RemoveBuff,
      abilityID: event.abilityGameID,
      targetPlayerID,
    };
  }

  return null;
};

export type PersistedDungeonPull = Pick<
  Pull,
  "id" | "x" | "y" | "startTime" | "endTime" | "isWipe"
> & {
  maps: PullZone["zoneID"][];
  boundingBox: ReportMapBoundingBox;
  enemyNPCs: Pick<
    Required<DeepNonNullable<ReportDungeonPullNpc>>,
    "gameID" | "minimumInstanceID" | "maximumInstanceID" | "id"
  >[];
};

const getProcessorParams = (
  event: AnyEvent,
  pull: PersistedDungeonPull,
  actorPlayerMap: Map<number, number>
): Parameters<Processor<AnyEvent>>[1] => {
  const params: Parameters<Processor<AnyEvent>>[1] = {
    sourceNPCID: null,
    sourcePlayerID: null,
    targetNPCID: null,
    targetPlayerID: null,
  };

  if ("sourceID" in event) {
    params.sourcePlayerID = actorPlayerMap.get(event.sourceID) ?? null;
    params.sourceNPCID =
      pull.enemyNPCs.find((npc) => npc.id === event.sourceID)?.gameID ?? null;
  }

  if ("targetID" in event) {
    params.targetPlayerID = actorPlayerMap.get(event.targetID) ?? null;
    params.targetNPCID =
      pull.enemyNPCs.find((npc) => npc.id === event.targetID)?.gameID ?? null;
  }

  return params;
};

export const processEvents = (
  pull: PersistedDungeonPull,
  events: AnyEvent[],
  actorPlayerMap: Map<number, number>
): Omit<Prisma.EventCreateManyInput, "pullID">[] => {
  return events
    .map<Omit<Prisma.EventCreateManyInput, "pullID"> | null>((event) => {
      const params = getProcessorParams(event, pull, actorPlayerMap);

      switch (event.type) {
        case "applybuff":
          return applyBuffProcessor(event, params);
        case "applybuffstack":
          return applyBuffStackProcessor(event, params);
        case "applydebuff":
          return applyDebuffProcessor(event, params);
        case "begincast":
          return beginCastProcessor(event, params);
        case "cast":
          return castProcessor(event, params);
        case "damage":
          return damageProcessor(event, params);
        case "death":
          return deathProcessor(event, { ...params, pull });
        case "heal":
          return healProcessor(event, params);
        case "interrupt":
          return interruptProcessor(event, params);
        case "removebuff":
          return removeBuffProcessor(event, params);
        default:
          console.log(event);
          return null;
      }
    })
    .filter(
      (dataset): dataset is Omit<Prisma.EventCreateManyInput, "pullID"> =>
        dataset !== null
    );
};
