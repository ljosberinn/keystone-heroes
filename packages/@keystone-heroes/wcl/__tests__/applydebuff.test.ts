import { DOS_URN } from "@keystone-heroes/wcl/queries/events/dungeons/dos";

import { SOA_SPEAR } from "../src/queries/events/dungeons/soa";
import type { ApplyDebuffEvent } from "../src/queries/events/types";
import { applyDebuffProcessor } from "../src/transform/events/applydebuff";
import allEvents from "./fixtures/allEvents.json";

describe("applydebuff", () => {
  const params = {
    targetPlayerID: null,
    targetNPCID: null,
    sourcePlayerID: null,
    sourceNPCID: null,
  };

  const applyDebuffEvents = allEvents.filter(
    (event): event is ApplyDebuffEvent => event.type === "applydebuff"
  );

  test("tracks SOA Spear", () => {
    const result = applyDebuffEvents
      .filter((event) => event.abilityGameID === SOA_SPEAR)
      .map((event) => applyDebuffProcessor(event, params));

    expect(result).toMatchSnapshot();
  });

  test("tracks DOS Urn", () => {
    const result = applyDebuffEvents
      .filter((event) => event.abilityGameID === DOS_URN)
      .map((event) => applyDebuffProcessor(event, params));

    expect(result).toMatchSnapshot();
  });

  test("returns null for unknown applydebuff events", () => {
    expect(
      applyDebuffProcessor(
        {
          type: "applydebuff",
          abilityGameID: 1,
          sourceID: 1,
          targetID: 0,
          timestamp: 1,
        },
        params
      )
    ).toBeNull();
  });
});
