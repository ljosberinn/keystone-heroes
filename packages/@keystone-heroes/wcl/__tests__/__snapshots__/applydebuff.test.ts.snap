// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applydebuff tracks DOS Urn 1`] = `
Array [
  Object {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "timestamp": 647832,
  },
  Object {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "timestamp": 1286270,
  },
  Object {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "timestamp": 1286270,
  },
  Object {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "timestamp": 1286270,
  },
  Object {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "timestamp": 1286270,
  },
  Object {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "timestamp": 1286270,
  },
  Object {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "timestamp": 1286270,
  },
]
`;

exports[`applydebuff tracks MOTS maze failures 1`] = `
Array [
  Object {
    "abilityID": 323881,
    "eventType": "ApplyDebuff",
    "timestamp": 35495540,
  },
  Object {
    "abilityID": 323881,
    "eventType": "ApplyDebuff",
    "timestamp": 35495540,
  },
  Object {
    "abilityID": 323881,
    "eventType": "ApplyDebuff",
    "timestamp": 35495540,
  },
  Object {
    "abilityID": 323881,
    "eventType": "ApplyDebuff",
    "timestamp": 35495540,
  },
  Object {
    "abilityID": 323881,
    "eventType": "ApplyDebuff",
    "timestamp": 35495540,
  },
]
`;

exports[`applydebuff tracks SOA Spear 1`] = `
Array [
  Object {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "timestamp": 11913707,
  },
  Object {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "timestamp": 11913707,
  },
  Object {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "timestamp": 11913707,
  },
  Object {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "timestamp": 11913707,
  },
  Object {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "timestamp": 11913707,
  },
  Object {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "timestamp": 11913707,
  },
  Object {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "timestamp": 11913707,
  },
  Object {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "timestamp": 11913707,
  },
  Object {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "timestamp": 11913879,
  },
  Object {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "timestamp": 12840160,
  },
]
`;
